### PR TITLE
some updates to code metadata handling

### PR DESCRIPTION
**Pull request by JeffBezanson**
- detect pure and inline meta once, in the front end
- remove `skip_meta` since it probably assumes too much about where meta nodes occur
- add jl_ prefix to has_meta
- remove location/pure/inline meta early, in the front end
  (the first line node isn't needed, since this info is in the Method object)
- remove null_sym; this Expr head doesn't occur in the IR

### CODE REVIEW

File: src/alloc.c [hunk 1]
```diff
@@ -315,8 +315,28 @@ static void jl_lambda_info_set_ast(jl_lambda_info_t *li, jl_expr_t *ast)
...
+            size_t k, ins = 0, na = jl_expr_nargs(st);
+            jl_value_t **meta = (jl_value_t**)jl_array_data(((jl_expr_t*)st)->args);
+            for(k=0; k < na; k++) {
+                jl_value_t *ma = meta[k];
+                if (ma == (jl_value_t*)pure_sym)
+                    li->pure = 1;
+                else if (ma == (jl_value_t*)inline_sym)
+                    li->inlineable = 1;
+                else
+                    meta[ins++] = ma;
```
Comment (vtjnash) [inline src/alloc.c]: A wb might be needed here if we do array carding?

File: test/reflection.jl [hunk 2]
```diff
@@ -364,9 +364,7 @@ function test_typed_ast_printing(f::ANY, types::ANY, must_used_vars)
     for str in (sprint(io->code_warntype(io, f, types)),
                 sprint(io->show(io, li)))
         # Test to make sure the clearing of file path below works
-        # If we don't store the full path in line number node/ast printing
-        # anymore, the test and the string replace below should be fixed.
-        @test contains(str, @__FILE__)
+        @test string(li.def.file) == @__FILE__
```
Comment (vtjnash) [inline test/reflection.jl]: Why did this test change but not the line after it (which is what this test and the deleted comment were here to protect)

File: src/codegen.cpp [hunk 3]
```diff
@@ -4723,7 +4719,7 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
                                              0,
                                              true,
                                              nullptr);
-                builder.SetCurrentDebugLocation(DebugLoc::get(0, 0, (MDNode*)SP, builder.getCurrentDebugLocation()));
+                builder.SetCurrentDebugLocation(DebugLoc::get(inlfline, 0, (MDNode*)SP, builder.getCurrentDebugLocation()));
             }
```
Comment (vtjnash) [inline src/codegen.cpp]: Typically, the first statements emitted for a function should have a line number of zero. I'm not sure that ends up mattering much right now for us  (due to the way we do inlining), but normally it helps debuggers avoid stopping in uninitiallized stack frames and profilers to  properly account for the cost of the prologue vs. the function body.

File: src/codegen.cpp [hunk 4]
```diff
@@ -4707,10 +4693,20 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
                 DI_sp_stack.push_back(SP);
                 DI_loc_stack.push_back(builder.getCurrentDebugLocation());
                 std::string inl_name;
-                if (jl_array_len(stmt_e->args) > 2)
-                    inl_name = jl_symbol_name((jl_sym_t*)jl_exprarg(stmt_e, 2));
-                else
+                int inlfline = 0;
```
Comment (vtjnash) [inline src/codegen.cpp]: This variable name is unreadable

File: src/codegen.cpp (hunk 3 above)
Comment (JeffBezanson) [inline src/codegen.cpp]: I don't understand --- surely we don't want the line of the first statement to actually be reported as zero. Should I do something like calling `SetCurrentDebugLocation` with 0, followed immediately by calling it again with the actual number?

Comment (vtjnash) [inline src/codegen.cpp]: I don't think we have any intermediate instructions to hide, so it's fine to fold those two calls (the DebugLoc set in the builder is only side-effecting on the instructions inserted by that builder). Just noting that if we want to do work to create a call frame or such, those instructions should report a line of 0 to the debugger.

**Pull request merged**